### PR TITLE
fromJSON and toJSON

### PR DESCRIPTION
**Pull request by mischnic**
There are rather ugly things in here, but let me know what you think...

### CODE REVIEW

File: lib/input.js [hunk 1]
```diff
@@ -49,42 +49,41 @@ class Input {
...
+      for (let i = 0, l = lines.length; i < l; i++) {
+        this.lineToIndex[i] = prevIndex
+        prevIndex += lines[i].length + 1
+      }
 
-    for (let i = 0, l = lines.length; i < l; i++) {
-      lineToIndex[i] = prevIndex
-      prevIndex += lines[i].length + 1
+      this.lastLine = this.lineToIndex[this.lineToIndex.length - 1]
     }
```
Comment (mischnic) [inline lib/input.js:63]: This change ensures that it doesn't try to store state in a function, which can't be serialized into JSON. This seems cleaner to me anyway (though an alternative would be to exclude this function from being returned in `Input#toJSON()`)

Comment (ai) [reply lib/input.js:63]: Yeap. Removing that method replacing hack can be nice.

But I afraid that `this.lineToIndex` will bloat up `console.log(node)`. What if we will use `let lineToIndecCache = new Symbol()`?

File: test/fromJSON.test.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      stringifier: postcss.stringify
+    }
+  ).css
+
+  let root = postcss.parse(cssWithMap)
+
+  let json = root.toJSON()
+  let serialized = v8.serialize(json)
+  let deserialized = v8.deserialize(serialized)
+  // @ts-ignore
```
Comment (ai) [inline test/fromJSON.test.ts:22]: Am I right that you are planning to add types later?

File: test/fromJSON.test.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+it('can rehydrate a JSON AST', () => {
+  let cssWithMap = postcss().process(
+    '.foo { color: red; font-size: 12pt; } /* abc */ @media (width: 60em) { }',
+    {
+      from: undefined,
+      map: {
+        inline: true
+      },
+      stringifier: postcss.stringify
```
Comment (ai) [inline test/fromJSON.test.ts:13]: `postcss.stringify` is a default stringifier. I think you can remove it.

File: test/fromJSON.test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import v8 from 'v8'
+
+import postcss, { Root } from '../lib/postcss.js'
+
+it('can rehydrate a JSON AST', () => {
```
Comment (ai) [inline test/fromJSON.test.ts:5]:
```suggestion
it('rehydrates a JSON AST', () => {
```

File: lib/postcss.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ let AtRule = require('./at-rule')
 let Result = require('./result.js')
 let Input = require('./input')
 let parse = require('./parse')
+let fromJSON = require('./fromJSON')
```
Comment (ai) [inline lib/postcss.js:15]: Can you keep imports order based on variable length? (I know that it is a strange system 😅)

File: lib/input.js [hunk 6]
```diff
@@ -168,6 +167,21 @@ class Input {
   get from () {
     return this.file || this.id
   }
+
+  toJSON () {
+    let json = {}
+    if (this.hasBOM != null) json.hasBOM = this.hasBOM
```
Comment (ai) [inline lib/input.js:173]: It will be nice to iterate throw through `['hasBOM', 'css', 'file', 'id']`

File: lib/fromJSON.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+let Declaration = require('./declaration')
+let Comment = require('./comment')
+let AtRule = require('./at-rule')
+let Root = require('./root')
+let Rule = require('./rule')
+let Input = require('./input')
+let PreviousMap = require('./previous-map')
+
+function fromJSON (json) {
```
Comment (ai) [inline lib/fromJSON.js:11]: We will need `fromJSON.d.ts` file

File: lib/fromJSON.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  } else if (json.type === 'decl') {
+    return new Declaration(defaults)
+  } else if (json.type === 'rule') {
+    return new Rule(defaults)
+  } else if (json.type === 'comment') {
+    return new Comment(defaults)
+  } else if (json.type === 'atrule') {
+    return new AtRule(defaults)
+  } else {
+    throw new Error('Unknown node type ' + json.type)
```
Comment (ai) [inline lib/fromJSON.js:42]: We will need a test for this case too

**Comment by ai** [comment]
Looks good! 👍

Let’s finish small changes and release under 8.2

File: test/fromJSON.test.ts (hunk 3 above)
Comment (mischnic) [reply test/fromJSON.test.ts:13]: "You did not set any plugins, parser, or stringifier. Right now, PostCSS does nothing. Pick plugins for your case on https://www.postcss.parts/ and use them in postcss.config.js."

File: test/fromJSON.test.ts (hunk 2 above)
Comment (mischnic) [reply test/fromJSON.test.ts:22]: The return value of `fromJSON` depends on the parameter's node type. I've instead casted the type with `as Root`, is that acceptable?

File: test/fromJSON.test.ts (hunk 3 above)
Comment (ai) [reply test/fromJSON.test.ts:13]: Try `postcss([])`

File: test/fromJSON.test.ts (hunk 2 above)
Comment (ai) [reply test/fromJSON.test.ts:22]: `as Root` :+1:

File: lib/fromJSON.js (hunk 7 above)
Comment (mischnic) [reply lib/fromJSON.js:11]: Should they be in that file or in `postcss.d.ts`? parse, stringify, ... are also documented in the `postcss.d.ts`

Comment (ai) [reply lib/fromJSON.js:11]: Yeap, we can do it in `postcss.d.ts` too.

File: test/fromJSON.test.ts (hunk 3 above)
Comment (mischnic) [reply test/fromJSON.test.ts:13]: This also prints that warning:
```js
  let cssWithMap = postcss([]).process(
    '.foo { color: red; font-size: 12pt; } /* abc */ @media (width: 60em) { }',
    {
      map: {
        inline: true
      },
    }
  ).css
```

Comment (ai) [reply test/fromJSON.test.ts:13]: OK, let’s use the current code

**Comment by ai** [comment]
I will release 8.2 on the next week (I need to promote previous 8.1.14 release first since it solved some painful 8.x migration problems)

**Pull request merged**